### PR TITLE
Remoted tcp udp secure connection handler

### LINKED ISSUE #7371: TCP and UDP connection handling in remoted
Hi team,

To implement the feature described in the issue `Support TCP and UDP for remoting at the same time` #1838 it is necessary to adapt the `HandleRemote()` and `HandleSecure()` functions to be able to handle TCP and/or UDP connections simultaneously, if required.

In order to handle both types of protocols simultaneously, in a single thread, it is necessary to:
- Adapt the code, so that one or both protocols can bind their corresponding ports in `HandleRemote()`.
- Add the UDP socket to the watching list of file descriptors to be monitored by `w_notify()` in `HandleSecure()`.

Regards.

### LINKED ISSUE #7373: Response from Remoted according to protocol received
Hi team,

To implement the feature described in the issue `Support TCP and UDP for remoted at the same time` #1838 it is necessary to adapt the  code to be able to answer to TCP or UDP messages, according to the protocol that the agent is using.

In order to do so, it is necessary to:
- Detect and store the protocol that the agent is using (`HandleSecure()` / `HandleSecureMessage()` / `send_file_toagent()`). 
- Handle the response accordingly given the stored agent protocol (`send_msg()`)

Regards.

### PR DESCRIPTION
**Pull request by juliancnn**
|Related issue|
|---|
|Closes #7371|
|Closes #7373|

## Description

Hi Team!,

While diving into the issues #7371 and #7373 we realized that these are deeply related, that is why we decided to close both issues in a single PR.
This PR is the final stage in order to implement the feature described in the issue: **Support TCP and UDP for remote at the same time #1838**. 

This PR allows supporting TCP and/or UDP protocols on the following features:

- Manager active check (`#Ping`  `#Pong`).
- Agent connection and logs sending
- Active Response 
- API requests-related interaction between manager and agents
- Shared files
- Centralized configuration (Agents can be configured remotely by using the **agent.conf** file.)
- Get configuration in JSON format

### Feature Tests: With TCP protocol 

`<protocol>tcp</protocol>`

- [ ] `#Ping`  `#Pong` (i.e. `go run ./wazuh_sock.go -s 192.168.0.35:1514 -p tcp `)
- [ ] Agent connect and send logs
- [ ] Active Response 
- [ ] Remote API requeset (i.e.  `go run ./wazuh_sock.go -f /var/ossec/queue/ossec/request -m "019 logcollector getstate"`)
- [ ] Shared files
- [ ] Centralized configuration (Agents can be configured remotely by using the **agent.conf** file.)
- [ ]  Get configuration in JSON format (`go run ./wazuh_sock.go -f /var/ossec/queue/ossec/request -m "getconfig remote"`)
- [ ] WPK

### Feature Tests: With UDP protocol 

`<protocol>udp</protocol>`

- [ ] `#Ping`  `#Pong` (i.e `go run ./wazuh_sock.go -s 192.168.0.35:1514 -p udp -h=false`)
- [ ] Agent connect and send logs
- [ ] Active Response 
- [ ] Remote API requeset (i.e.  `./secure_tcp.go -f /var/ossec/queue/ossec/request -m "019 agent getstate"`)
- [ ] Shared files
- [ ] Centralized configuration (Agents can be configured remotely by using the **agent.conf** file.)
- [ ]  Get configuration in JSON format (`go run ./wazuh_sock.go -f /var/ossec/queue/ossec/request -m "getconfig remote"`)
- [ ] WPK

### Feature Tests: With TCP + UDP

`<protocol>udp, tcp</protocol>`

- [ ] `#Ping`  `#Pong`
- [ ] Agent connect and send logs
- [ ] Active Response 
- [ ] Remote API requeset (i.e.  `./secure_tcp.go -f /var/ossec/queue/ossec/request -m "019 agent getstate"`)
- [ ] Shared files
- [ ] Centralized configuration (Agents can be configured remotely by using the **agent.conf** file.)
- [ ]  Get configuration in JSON format (`go run ./wazuh_sock.go -f /var/ossec/queue/ossec/request -m "getconfig remote"`)
- [ ] WPK

Regards,
Julian

## Tests

- Compilation without warnings in every supported platform
  - [ ] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [ ] Source installation
- [ ] Package installation
- [ ] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for Windows
  - [ ] Scan-build report
  - [ ] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Review by nmkoremblum — APPROVED** [review]
LGTM thanks! :D

File: src/config/remote-config.c [hunk 1]
```diff
@@ -296,10 +296,10 @@ STATIC int w_remoted_get_net_protocol(const char * content) {
             char * word = &(proto_arr[current])[strspn(proto_arr[current], " ")];
             word[strcspn(word, " ")] = '\0';
 
-            if (strcasecmp(word, REMOTED_PROTO_TCP_STR) == 0) {
-                retval |= REMOTED_PROTO_TCP;
-            } else if(strcasecmp(word, REMOTED_PROTO_UDP_STR) == 0) {
-                retval |= REMOTED_PROTO_UDP;
+            if (strcasecmp(word, REMOTED_NET_PROTOCOL_TCP_STR) == 0) {
+                retval |= REMOTED_NET_PROTOCOL_TCP;
+            } else if(strcasecmp(word, REMOTED_NET_PROTOCOL_UDP_STR) == 0) {
```
Comment (jnasselle) [inline src/config/remote-config.c:301]: Add space between `if` and parenthesis

File: src/remoted/manager.c [hunk 2]
```diff
@@ -933,6 +934,15 @@ int send_file_toagent(const char *agent_id, const char *group, const char *name,
         return (-1);
     }
 
+    /* The following code is used to get the protocol that the client is using in order to answer accordingly */
+    key_lock_read();
+    protocol = w_get_agent_net_protocol_from_keystore(&keys, agent_id);
+    key_unlock();
+    if (protocol < 0) {
+        merror(AR_NOAGENT_ERROR, agent_id);
+        return (-1);
```
Comment (jnasselle) [inline src/remoted/manager.c:943]: Remove parenthesis for return value

File: src/config/remote-config.h [hunk 3]
```diff
@@ -14,19 +14,21 @@
...
+#define REMOTED_NET_PROTOCOL_TCP_STR "TCP" ///< String to represent the TCP protocol
+#define REMOTED_NET_PROTOCOL_UDP_STR "UDP" ///< String to represent the UDP protocol
+#define REMOTED_NET_PROTOCOL_DEFAULT_STR  (REMOTED_NET_PROTOCOL_DEFAULT == REMOTED_NET_PROTOCOL_TCP \
+                ? REMOTED_NET_PROTOCOL_TCP_STR : REMOTED_NET_PROTOCOL_UDP_STR) ///< String to represent default protocol
 
-#define REMOTED_PROTO_TCP_STR "TCP"
-#define REMOTED_PROTO_UDP_STR "UDP"
-#define REMOTED_PROTO_DEFAULT_STR  (REMOTED_PROTO_DEFAULT == REMOTED_PROTO_TCP \
-                              ? REMOTED_PROTO_TCP_STR : REMOTED_PROTO_UDP_STR) ///< String to represent default protocol
+#define REMOTED_NET_PROTOCOL_TCP_STR "TCP"
```
Comment (JcabreraC) [inline src/config/remote-config.h:26]: The following defines are duplicated:
- REMOTED_PROTO_TCP_STR
- REMOTED_PROTO_UDP_STR
- REMOTED_PROTO_DEFAULT_STR

**Review by JcabreraC — CHANGES_REQUESTED** [review]

**Review by jnasselle — APPROVED** [review]

**Review by JcabreraC — APPROVED** [review]

**Pull request merged**